### ISSUE TITLE
Refactor away `composer.core`

### ISSUE THREAD

**Issue by ravi-mosaicml** [mosaicml/composer#429]
Refactoring away `composer.core` will help organize the imports (and thus API reference). For example, the logging base class is in composer.core, but the logger hparams base and logger implementations  are in composer.loggers. This complicates the imports, leads to circular dependencies, and reduces the docs readability (since now one must look in two different places for the docs).

Instead, would it be possible to refactor away `composer.core`? I am thinking of a refactor like the following:
- [ ] `composer.core.algorithm` -> `composer.algorithms.base`
- [ ] `composer.core.callback` -> `composer.callbacks.base`
- [ ] `composer.core.data_spec` -> `composer.datasets.data_spec`
- [ ] `composer.core.engine` -> `composer.engine.engine` (new module for `composer.engine`)
- [ ] `composer.core.evaluator` -> `composer.trainer.evaluator` (depends on #329)
- [ ] `composer.core.event` -> `composer.engine.event` (new module for `composer.engine`)
- [ ] `composer.core.precision` -> `composer.types.precision` (new module for `composer.types`)
- [x] `composer.core.profiler` -> `composer.profiler.profiler` (addressed in #422)
- [ ] `composer.core.serializable` -> `composer.types.serializable`  (new module for `composer.types`)
- [ ] `composer.core.state` -> `composer.engine.state` (new module for `composer.engine`)
- [x] `composer.core.surgery` -> `composer.utils.surgery`
- [ ] `composer.core.time` -> `composer.time`
- [ ] `composer.core.types` -> `composer.types.types` (new module for `composer.types`)
- [x] `composer.core.logging` -> `composer.logging`

Because of integrated `core` is into the best of the codebase, it would probably be best to move one file per PR.

CC @jbloxham @anisehsani @bandish-shah @ajaysaini725

**Comment by jbloxham** [comment]
For refactors like this, I find it useful to think about import hierarchies of modules.

I think the new hierarchy after this change would look something like:

`types > utils > [datasets | models | optim] > engine > [algorithms | callbacks > loggers] > trainer`

The big flag I see here is the interaction between `algorithms`/`callbacks` and `engine`. The algorithm API depends on `engine.State`, but algorithms themselves are also in the state. Could we just remove algorithms and callbacks from the state?

**Comment by ravi-mosaicml** [comment]
> The big flag I see here is the interaction between `algorithms`/`callbacks` and `engine`. The algorithm API depends on `engine.State`, but algorithms themselves are also in the state. Could we just remove algorithms and callbacks from the state?

Yes! See https://github.com/mosaicml/composer/issues/155

**Comment by jbloxham** [comment]
Amazing! I'm all good with these changes then.

**Comment by hanlint** [comment]
A few other suggestions/thoughts:
* turn folders -> _.py files whenever we don't actually need them 
* should we be putting modules that are really tied to our trainer within the trainer module? 

r`state.algorithms`: do we expect algorithms to want to inspect what other algorithms are being applied and adjust their logic? or do we plan one monolithic `engine` that handles all this complexity?

**Comment by jbloxham** [comment]
I'm in favor of a monolithic `engine` for handling algorithm conflicts. If algorithm conflicts grow roughly with `O(n^2)`, we'll eventually have to do something smarter than manual fixes for every individual conflict anyways, in order to handle this problem scalably.

**Comment by jbloxham** [comment]
I think the `engine` module is also necessary because there is trainer-specific code that depends on algorithms, but also algorithms depend on some other trainer-specific code. The distinction between these two (thankfully disjoint?) bodies of code is the `engine` module and the `trainer` module.

**Comment by ravi-mosaicml** [comment]
> Yes I would love to move away from `core`! A few other suggestions/thoughts:
> 
> * turn folders -> _.py files whenever we don't actually need them

Sure! Makes sense...a folder with all private members and an `__init__.py` that rexports is equivalent to a file. 

> * should we be putting modules that are really tied to our trainer within the trainer module?

I presume yes? But which ones are tied to our trainer?

> `state.algorithms`: do we expect algorithms to want to inspect what other algorithms are being applied and adjust their logic? or do we plan one monolithic `engine` that handles all this complexity?

I would agree with Jamie here and say that we should have an engine coordinate this. But I could see an algorithm needing to provide additional metadata that the engine uses (not sure what this metadata would entail given our current set of algorithms). 

> I don't completely follow some of the proposed renaming above (e.g. `engine.events`, when events are shared with callbacks and the trainer loop itself?), but let's talk through them!

Agreed, I think it's least clear  `event`, `engine`, and `state` should live. I think perhaps this could remain in a much-slimmed-down core, or each could have its top-level module (or file)

**Comment by mvpatel2000** [comment]
Closing -- moved to JIRA and some of this we won't do
